accepted/0000-workspaces-part2.md

```diff
@@ -0,0 +1,196 @@
+
+Only a subset of commands are to be supported:
+
+- `fund` List funding info for all **workspaces**
+- `ls` List all packages including **workspaces**
+- `outdated` List outdated **dependencies** including **workspaces** and its **dependencies**
+- `run-script` Run arbitrary **scripts** in all **workspaces**, skip any **workspace** that does not have a targetting **script**
+- `rebuild` Rebuild all **workspaces**
+- `restart`
+- `start`
+- `stop`
+- `test` Run test **scripts** in all **workspaces**
+- `update` Updates a **dependency** across the entire installation tree, including **workspaces**
+- `view` View registry info, also including **workspaces**
+
+#### Test example:
+
+```
+├── package.json { "name": "foo", "workspaces": ["dep-a", "dep-b"] }
+├── dep-a
+│   └── package.json { "version": "1.0.0" }
+└── dep-b
+    └── package.json { "version": "1.3.1" }
+
+$ npm test
+
+> foo@1.0.0 test /Users/username/foo
+> echo "Error: no test specified" && exit 1
+
+Error: no test specified
+npm ERR! Test failed.  See above for more details.
+
+> dep-a@1.0.0 test /Users/username/foo/dep-a
+> done
+
+> dep-b@1.3.1 test /Users/username/foo/dep-b
+> done
+```
+
+**NOTE:** `publish` and `version` are very special cases in a **npm workspaces** setup and are outside of the scope of this RFC. These features will live in userland in the form of popular modules such as [Lerna](https://lerna.js.org/) for the foreseeable future.
+
+### 2. Filter a subset of workspaces
+
+Filter is done via positional argument and here are some of the reasons why we decided to go that route:
+- [Lerna filters](https://www.npmjs.com/package/@lerna/filter-options) were the starting point but early in the discussion some other considerations were brought forward
+- npm configs have some baggage, configs might apply to the larger behavior of the cli or to a specific subcommand and they're all mixed together
+- defining a subset of workspaces in a `.npmrc` file might be just a footgun leading users into confusion
+- Yarn v1 uses positional arguments for filter in the form of `yarn workspace <workspace_name> <cmd>`, so there's precedent
+
+Leading this RFC to a proposal using positional arguments. Rather than defining a strict UX, the current proposal is to adhere to a `workspace` command prefix in the form of either: `npm workspace <workspace_name> <command>` OR `npm ws <workspace_name> <command>` OR an even shorter shortcut using a symbol `npm :<workspace_name> <command>`, for example running `test` in a specific **workspace** named `foo` with these different ideas:
+
+- `npm workspace foo test`
+- `npm ws foo test`
+- `npm :foo test`
+
+Leaving space to gather feedback from the community on whether to introduce one of this proposed commands or even introducing some novel interface.
```

Comment:
I don't know if there are other examples of such syntax outside of Yarn. For me, a flag would look more usual. It can be something short like `npm -w foo test`. If you want to run tests in foo and bar, you could do either `npm -w foo -w bar test` or `npm -w foo,bar test`

Reply:
I really like using a shorter alias such as `-w` but the reason to steer away from named configs has more to do with the [specific **npm** behavior](https://github.com/npm/rfcs/pull/117/files#diff-828bb29c16ce81ff3a4108b6e8dab945R78)

Reply:
can't npm treat the `--workspace` flag differently? Not reading it from the rc file?

Reply:
this is related to the conversation going on at https://github.com/npm/rfcs/pull/133#issuecomment-621397885 and honestly we might just spin off an RFC/discussion around a new config management system for the cli

Reply:
thanks @zkochan we ended up going with that option! 😄